src/mscorlib/src/System/Buffer.cs

```diff
@@ -256,325 +256,162 @@ internal unsafe static void Memcpy(byte* dest, byte* src, int len)
-                    return;
-                case 14:
-#if BIT64
-                    *(long*)dest = *(long*)src;
-#else
-                *(int*)dest = *(int*)src;
-                *(int*)(dest + 4) = *(int*)(src + 4);
-#endif
-                    *(int*)(dest + 8) = *(int*)(src + 8);
-                    *(short*)(dest + 12) = *(short*)(src + 12);
-                    return;
-                case 15:
-#if BIT64
-                    *(long*)dest = *(long*)src;
-#else
-                *(int*)dest = *(int*)src;
-                *(int*)(dest + 4) = *(int*)(src + 4);
-#endif
-                    *(int*)(dest + 8) = *(int*)(src + 8);
-                    *(short*)(dest + 12) = *(short*)(src + 12);
-                    *(dest + 14) = *(src + 14);
-                    return;
-                case 16:
-#if BIT64
-                    *(long*)dest = *(long*)src;
-                    *(long*)(dest + 8) = *(long*)(src + 8);
+            const nuint CopyThreshold = 1024;
+#endif // PLATFORM_UNIX
 #else
-                *(int*)dest = *(int*)src;
-                *(int*)(dest + 4) = *(int*)(src + 4);
-                *(int*)(dest + 8) = *(int*)(src + 8);
-                *(int*)(dest + 12) = *(int*)(src + 12);
-#endif
-                    return;
-                case 17:
-#if BIT64
-                    *(long*)dest = *(long*)src;
-                    *(long*)(dest + 8) = *(long*)(src + 8);
+            const nuint CopyThreshold = 256;
+#endif // AMD64
+
+            // FCALL/PInvoke to the native version when the copy length exceeds the threshold or when the buffers are overlapping. This 
+            // check can produce false positives for lengths greater than Int32.MaxInt. But, this is fine because we want to FCALL/PInvoke 
+            // for large lengths anyways.
+            
+            if ((len > CopyThreshold) || ((nuint)dest < (nuint)srcEnd)) 
+            {
+#if AMD64 && !PLATFORM_UNIX
+                // FCall to JIT_MemCpy on x64 Windows
+                goto FCopy;
 #else
-                *(int*)dest = *(int*)src;
-                *(int*)(dest + 4) = *(int*)(src + 4);
-                *(int*)(dest + 8) = *(int*)(src + 8);
-                *(int*)(dest + 12) = *(int*)(src + 12);
-#endif
-                    *(dest + 16) = *(src + 16);
-                    return;
-                case 18:
-#if BIT64
-                    *(long*)dest = *(long*)src;
-                    *(long*)(dest + 8) = *(long*)(src + 8);
+                // PInvoke to CRT memmove for overlap checking and for avoiding the possibility of GC starvation and improper exception 
+                // handling (See comments for PR #6638). 
+                goto PInvoke;
+#endif // AMD64 && !PLATFORM_UNIX                
+            }
+            
+            byte* destEnd = dest + len;            
+            if (len <= 16) goto MCPY02;
+            if (len > 64) goto MCPY05;
+            
+            MCPY00:
+            // len > 16 && len <= 64
```

Comment:
Nit: Could be a `Debug.Assert` instead of a comment (same for other comments like this)